pkg/metrics/middleware.go

```diff
@@ -24,13 +28,48 @@ import (
 // around API calls.
 // It records the timestamp of an API call in the provided gauge.
 type APIEventTSHelper struct {
-	Next    http.Handler
-	TSGauge prometheus.Gauge
+	Next      http.Handler
+	TSGauge   prometheus.Gauge
+	Histogram *prometheus.HistogramVec
+}
+
+type responderWrapper struct {
+	http.ResponseWriter
+	code int
+}
+
+func (rw *responderWrapper) WriteHeader(code int) {
+	rw.code = code
+	rw.ResponseWriter.WriteHeader(code)
+}
+
+// getShortPath returns the API path trimmed after the 3rd slash.
+// examples:
+//  "/v1/config" -> "/v1/config"
+//  "/v1/endpoint/cilium-local:0" -> "/v1/endpoint"
+//  "/v1/endpoint/container-id:597.." -> "/v1/endpoint"
+func getShortPath(s string) string {
+	var idxSum int
+	for nThSlash := 0; nThSlash < 3; nThSlash++ {
+		idx := strings.IndexByte(s[idxSum:], '/')
+		if idx == -1 {
+			return s
+		}
+		idxSum += idx + 1
+	}
+	return s[:idxSum-1]
 }
 
 // ServeHTTP implements the http.Handler interface. It records the timestamp
 // this API call began at, then chains to the next handler.
 func (m *APIEventTSHelper) ServeHTTP(r http.ResponseWriter, req *http.Request) {
 	m.TSGauge.SetToCurrentTime()
-	m.Next.ServeHTTP(r, req)
+	duration := spanstat.Start()
+	rw := &responderWrapper{ResponseWriter: r}
+	m.Next.ServeHTTP(rw, req)
+	if req != nil && req.URL != nil && req.URL.Path != "" {
+		path := getShortPath(req.URL.Path)
+		took := float64(duration.End(true).Total().Seconds())
+		m.Histogram.WithLabelValues(path, req.Method, strconv.Itoa(rw.code)).Observe(took)
```

Comment:
I still think that the `response.code` is not the best variable to add here, for me will be true/false.

Reply:
@eloycoto Are you worried about too many labels? We can use `label_replace()` to turn it into a 4xx and 5xx label in the query function. I think it is valuable to know what type of error is being returned. We have well-documented error codes and a limited number thereof.

Reply:
Yep,not too worried about the number of labels, but yes, [0-5]xx works for me